### ISSUE TITLE
GPIO + Raspberry Pi + Service + Sudo

### ISSUE THREAD

**Issue by Belphemur** [node-red/node-red#568]
It seems Node-RED is trying to execute at load the nrgpio.py with root right using sudo.
The user that launch my Node-RED service is not in the sudoers and I don't plan to add it to it.

This is the error I'm getting:

```
Feb 16 17:34:18 raspy sudo[27872]: node-red : user NOT in sudoers ; TTY=unknown ; PWD=/home/pi/node-red-git ; USER=root ; COMMAND=/usr/bin/python -u /home/pi/node-red-git/nodes/core/hardware/nrgpio.py rev 0
```

I think this is for @dceejay :)

**Comment by dceejay** [comment]
Hi

The RPi.GPIO library uses access to /dev/mem which requires root. The only ways round this are to use a compiled program that you can assigned escalated privileges, or to run as sudo, or to add python to the allowable sudoers list of apps. 

Eg http://askubuntu.com/questions/39281/how-to-run-an-application-using-sudo-without-a-password

(The problem is that the access rights (ie root) get assigned dynamically to each pin as it is set as an input or output, etc, so can't be reassigned easily (eg at boot time) as that would have to be done by root anyway)

**Comment by Belphemur** [comment]
Hi Dave (Sorry for the lack of politeness, took me a good hour to find why the daemon stopped working after a pull), 
I understand, just it would be nice to have a warning in the shell script if the user doesn't have root privilege. Because like I reported, it made systemd kill Node-RED since it tried to access root.

I found this online: http://www.element14.com/community/thread/19995/l/anyone-know-how-to-access-raspi-gpio-without-sudo

Creating a program or script with the setuid bit to be sure it will always be run as root without the need for sudo.

**Comment by dceejay** [comment]
That was actually why I created the nrgpio script :-) So you can set that suid with chmod u+s .... May work.  But sadly pushing to git removes that flag (not un-reasonably)...  The Pi docs are due an update due to this and Pi2 issues so I will add some for both of these issues shortly.

**Comment by knolleary** [comment]
@dceejay can the script try to determine if it has the appropriate permissions before trying to sudo and log a more helpful message?

**Comment by dceejay** [comment]
Bien sur... But not until Friday.
(It should already log not executables and no access "nicely")

**Comment by Belphemur** [comment]
You could put the script with the setuid bit and executable bit in a tar and decompress it at the first load of the node. 

I still think it's something the user should be aware of since we're playing with root.  Like a warning in the console asking the user to do a set of command as root to set correctly the script. 

It would be in the installation guide on the website. 

On 16 February 2015 23:31:26 EET, Dave Conway-Jones <notifications@github.com> wrote:
>Bien sur... But not until Friday.
>(It should already log not executables and no access "nicely")
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/node-red/node-red/issues/568#issuecomment-74575419

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by dceejay** [comment]
Linux doesn't let you run scripts suid... so this is a non-starter. Currently the only option that definitely works is letting the user run python as sudo - via visudo.

**Comment by Belphemur** [comment]
Or using something else than python for it : http://www.element14.com/community/thread/19995/l/anyone-know-how-to-access-raspi-gpio-without-sudo?displayFullThread=true

**Comment by dceejay** [comment]
Well the whole point of moving to the python was to get the benefits of the RPi.GPIO library... but yes... 

I'm currently re-looking at the wiring-pi npm https://www.npmjs.com/package/wiring-pi as that now seems to be being maintained again... and has B+ support etc) but no mention of Pi 2 yet... so I'm not rushing to it just yet, as we have a stack ofother things on the go... But it does seem also to have some other goodness we could use such as servo support and so on... and hopefully not the permissions problem. (yet to be tested).

**Comment by maxnet** [comment]
>The RPi.GPIO library uses access to /dev/mem which requires root.

It can use ``/dev/gpiomem`` instead nowadays, which only requires that the user is part of the gpio group.

So I don't think there is any excuse left not to allow unprivileged users to use node-red, and I suggest:

1) removing the ``sudo`` at https://github.com/node-red/node-red/blob/master/nodes/core/hardware/nrgpio
2) modifying the startup script so that it does not use systemd to start node-red, but just calls ``node-red-pi``, if calling user is not a member of the sudo group, here: https://github.com/node-red/raspbian-deb-package/blob/master/resources/node-red-start#L51

**Comment by dceejay** [comment]
That set of nodes also talks to mouse buttons and keyboard codes... need to check they all would work without root also. 
Also why remove from systemd ?  That is the preferred way to auto start things at boot these days isn't it ? Systemd can still start things as a non-privileged user. Starting it that way means the logs go to the same place, and allow you to ctrl-c out of viewing the log without killing Node-RED.

**Comment by maxnet** [comment]
>That set of nodes also talks to mouse buttons and keyboard codes... need to check they all would work 
>without root also.

I think you currently access /dev/input/x evdev devices directly, and while that does not require root, it does require that the user is a member of ``input`` group.
Granting a user access to that on a multi-user system is a potential security risk as well, as it could allow them to keylog other users.

I think you can capture key events without special privileges through using XINPUT extension instead, if node-red is running under X.
But haven't tried it on the Raspberry platform.

>Also why remove from systemd ? That is the preferred way to auto start things at boot these days isn't it ? 

Not everyone wants to auto start things at boot.
My use-case involves multi-user environments, in which users do should be able to start node-red in their own session to play with it.
But it should be killed when they log out.

**Comment by dceejay** [comment]
will look at keyboard - but the requirement of course is that it is still able to grab keys when the app doesn't have focus - ie is run in background... and not necessarily under X, so may not be possible.
You can of course delete/ remove the gpio nodes if you don't need them - The alternative node-red-node-pi-gpiod uses the pigpiod daemon instead (which itself runs as root - but then the user code (node-red) doesn't need root access.

You may also delete / remove the node-red-start script and systemd files as part of your system setup/config - and just call node-red-pi or node-red if you prefer... so the user can just run it locally only. Up to you.

**Comment by maxnet** [comment]
>but the requirement of course is that it is still able to grab keys when the app doesn't have focus

That should be possible through the XINPUT extension.

>and not necessarily under X

Well, you could always fallback to what you are currently using if not running under X (easy to check, if running under X "DISPLAY" environment variable is set).

>The alternative node-red-node-pi-gpiod uses the pigpiod daemon instead (which itself runs as root - but 
>then the user code (node-red) doesn't need root access.

I am not against making normal users part of the ``gpio`` group.
No need to sudo in that case, even with RPi.GPIO.

>You may also delete / remove the node-red-start script and systemd files as part of your system 
>setup/config - and just call node-red-pi or node-red if you prefer... so the user can just run it locally only. 
>Up to you.

Would prefer to see this fixed upstream, rather than me fixing things locally.
Your Raspbian package chooses to install node-red globally (in ``/usr/share/applications/Node-RED.desktop``) for every user, and not just for privileged users (e.g. pi)
In those circumstances I do not think it is unreasonable to expect that some effort is taken to also make it possible that normal users are able to use the icon you put in their menu...
Instead of it erroring out, like it does now.

If you prefer to use systemd if possible, you could use the standard ``groups`` shell command to check group membership, and have a construct like this in node-red-start:

```
if groups | grep -w sudo; then
    # Current user is member of sudo group, start node-red through systemd
    sudo service nodered start
    echo " "
    sudo journalctl -f -n 0 -u nodered -o cat  
else
    # Start node-red as unprivileged user
    node-red-pi
fi
```

**Comment by dceejay** [comment]
I can live with that.
Will remove sudo from the script, and change the start script to detect sudo or not.

**Comment by NeJamais** [comment]
I had a problem where node red could not operate the GPIO.  This was down to access privilleges on the /dev/gpiomem file.  I am running a model B upgraded from wheezy to jessie to current stretch.  Full details here [https://www.raspberrypi.org/forums/viewtopic.php?f=29&t=183481&p=1245876#p1245876](url)

**Comment by knolleary** [comment]
We don't use sudo anymore for these scripts.
